crates/services/p2p/src/dnsaddr_resolution.rs

```diff
@@ -0,0 +1,111 @@
+        &'a self,
+        addr: &'a str,
+        depth: usize,
+    ) -> Pin<
+        Box<dyn std::future::Future<Output = anyhow::Result<Vec<Multiaddr>>> + Send + 'a>,
+    > {
+        Box::pin(async move {
+            if depth >= MAX_DNS_LOOKUPS {
+                return Err(anyhow!("Maximum DNS lookup depth exceeded"));
+            }
+
+            let mut multiaddrs = vec![];
+            let dns_lookup_url = format!("{}{}", DNSADDR_PREFIX, addr);
+            let txt_records = self.resolver.txt_lookup(dns_lookup_url).await?;
+
+            for record in txt_records {
+                let parsed = record.to_string();
+                if !parsed.starts_with("dnsaddr") {
+                    continue;
+                }
+
+                let dnsaddr_value = parsed
+                    .split("dnsaddr=")
+                    .nth(1)
+                    .ok_or_else(|| anyhow!("Invalid DNS address: {:?}", parsed))?;
+
+                // Check if the parsed value is a multiaddress or another dnsaddr.
+                if dnsaddr_value.starts_with("/dnsaddr") {
+                    let nested_dnsaddr = dnsaddr_value
+                        .split('/')
+                        .nth(2)
+                        .ok_or_else(|| anyhow!("Invalid nested dnsaddr"))?;
+                    // Recursively resolve the nested dnsaddr
+                    #[allow(clippy::arithmetic_side_effects)]
+                    let nested_addrs =
+                        self.resolve_recursive(nested_dnsaddr, depth + 1).await?;
+                    multiaddrs.extend(nested_addrs);
+                } else if let Ok(multiaddr) = dnsaddr_value.parse::<Multiaddr>() {
+                    multiaddrs.push(multiaddr);
+                }
+            }
+
+            Ok(multiaddrs)
+        })
+    }
+}
+
+#[allow(non_snake_case)]
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::collections::HashSet;
+
+    #[tokio::test]
+    async fn dns_resolver__parses_all_multiaddresses_from_mainnet_dnsaddr_entry() {
+        // given
+        let resolver = DnsResolver::new().await.unwrap();
+        let expected_multiaddrs: HashSet<Multiaddr> = [
+            "/dns/p2p-mainnet.fuel.network/tcp/30336/p2p/16Uiu2HAkxjhwNYtwawWUexYn84MsrA9ivFWkNHmiF4hSieoNP7Jd",
+            "/dns/p2p-mainnet.fuel.network/tcp/30337/p2p/16Uiu2HAmQunK6Dd81BXh3rW2ZsszgviPgGMuHw39vv2XxbkuCfaw",
+            "/dns/p2p-mainnet.fuel.network/tcp/30333/p2p/16Uiu2HAkuiLZNrfecgDYHJZV5LoEtCXqqRCqHY3yLBqs4LQk8jJg",
+            "/dns/p2p-mainnet.fuel.network/tcp/30334/p2p/16Uiu2HAkzYNa6yMykppS1ij69mKoKjrZEr11oHGiM5Mpc8nKjVDM",
+            "/dns/p2p-mainnet.fuel.network/tcp/30335/p2p/16Uiu2HAm5yqpTv1QVk3SepUYzeKXTWMuE2VqMWHq5qQLPR2Udg6s"
+        ].iter().map(|s| s.parse().unwrap()).collect();
+
+        // when
+        // run a `dig +short txt _dnsaddr.mainnet.fuel.network` to get the TXT records
+        let multiaddrs = resolver
+            .lookup_dnsaddr("mainnet.fuel.network")
```

Comment:
Can we switch this to `core-test.fuellabs.net`? The records on the `mainnet.fuel.network` record can and will change over time, but I just setup the `core-test` DNS record to be static.

Reply:
Thanks for setting this up. I created [this follow-up issue](https://github.com/FuelLabs/fuel-core/issues/2212) to use the `core-test.fuellabs.net` hostname in the test, since this PR got auto-merged after my approval.